cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_train_dense_default_impl.i

```diff
@@ -905,6 +910,667 @@ void UnorderedRespHelper<algorithmFPType, cpu>::finalizeBestSplit(const IndexTyp
+
+    template <int K, bool noWeights>
+    int findBestSplitFewClasses(int nDiffFeatMax, size_t n, size_t nMinSplitPart, const ImpurityData & curImpurity, TSplitData & split,
+                                const algorithmFPType minWeightLeaf, const algorithmFPType totalWeights) const;
+
+    template <bool noWeights>
+    int findBestSplitFewClassesDispatch(int nDiffFeatMax, size_t n, size_t nMinSplitPart, const ImpurityData & curImpurity, TSplitData & split,
+                                        const algorithmFPType minWeightLeaf, const algorithmFPType totalWeights) const;
+
+private:
+    size_t nClasses() const { return this->_nClasses; }
+
+    template <bool noWeights>
+    bool findBestSplitOrderedFeature(const algorithmFPType * featureVal, const IndexType * aIdx, size_t n, size_t nMinSplitPart,
+                                     const algorithmFPType accuracy, const ImpurityData & curImpurity, TSplitData & split,
+                                     const algorithmFPType minWeightLeaf, const algorithmFPType totalWeights) const;
+
+    bool findBestSplitCategoricalFeature(const algorithmFPType * featureVal, const IndexType * aIdx, size_t n, size_t nMinSplitPart,
+                                         const algorithmFPType accuracy, const ImpurityData & curImpurity, TSplitData & split,
+                                         const algorithmFPType minWeightLeaf, const algorithmFPType totalWeights) const;
+};
+
+template <typename algorithmFPType, CpuType cpu>
+bool UnorderedRespHelperRandom<algorithmFPType, cpu>::findBestSplitForFeature(const algorithmFPType * featureVal, const IndexType * aIdx, size_t n,
+                                                                              size_t nMinSplitPart, const algorithmFPType accuracy,
+                                                                              const ImpurityData & curImpurity, TSplitData & split,
+                                                                              const algorithmFPType minWeightLeaf,
+                                                                              const algorithmFPType totalWeights) const
+{
+    const bool noWeights = !this->_weights;
+    if (noWeights)
+    {
+        return split.featureUnordered ?
+                   findBestSplitCategoricalFeature(featureVal, aIdx, n, nMinSplitPart, accuracy, curImpurity, split, minWeightLeaf, totalWeights) :
+                   findBestSplitOrderedFeature<true>(featureVal, aIdx, n, nMinSplitPart, accuracy, curImpurity, split, minWeightLeaf, totalWeights);
+    }
+    else
+    {
+        return split.featureUnordered ?
+                   findBestSplitCategoricalFeature(featureVal, aIdx, n, nMinSplitPart, accuracy, curImpurity, split, minWeightLeaf, totalWeights) :
+                   findBestSplitOrderedFeature<false>(featureVal, aIdx, n, nMinSplitPart, accuracy, curImpurity, split, minWeightLeaf, totalWeights);
+    }
+}
```

Comment:
Why do you need to redefine findBestSplitForFeature method (and some other methods as well)? Isn't it the same as in the base class?
It seems you can just reuse it via using-declaration.
Or maybe I am missing something.

Looks like almost all the implementation of UnorderedRespHelper class was copy-pasted. And the intension was that it could be inherited from the base class.

It

Reply:
I have re-worked the code to use the crtp way that you showed me. This now removes the code duplication.